File: src/main/java/akostenko/aicars/track/TrackSection.java
package akostenko.aicars.track;

import static java.lang.StrictMath.abs;
import static java.lang.StrictMath.signum;
import static java.lang.StrictMath.sin;

import akostenko.aicars.math.Polar;
import akostenko.aicars.math.Vector;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TrackSection {

    private static final int wayPointStep = 1; // m

    private final int indexOnTrack;
    private final Vector start; // starting point
    private final double length;
    private final double heading; // for turns it's a tangent line to the beginning point
    private final double radius;
    private final double width;
    private final List<TrackWayPoint> wayPoints;
    private final int distanceFromStart;

    TrackSection(int distanceFromStart, int indexOnTrack, Vector start, double length, double radius, double heading, double width) {
        if (length <= 0) {
            throw new IllegalArgumentException("Length of track section must be positive!");
        }

        this.distanceFromStart = distanceFromStart;
        this.indexOnTrack = indexOnTrack;
        this.start = start;
        this.length = length;
        this.radius = radius;
        this.heading = heading;
        this.width = width;

        int totalWayPoints = ((int) length / wayPointStep) + (length % wayPointStep > 0 ? 1 : 0);
        List<TrackWayPoint> wayPoints = new ArrayList<>(totalWayPoints);
        Vector wayPointPosition = start;
        double headingToNextWayPoint = heading;

        if (radius == 0) {
            for (int i = 0; i < totalWayPoints; i++) {
                wayPoints.add(new TrackWayPoint(this, wayPointPosition, i, distanceFromStart + i));
                wayPointPosition = wayPointPosition.plus(new Polar(wayPointStep, headingToNextWayPoint));
            }
        } else {
            double angleBetweenWayPoints = wayPointStep / abs(radius);
            double distanceBetweenWayPoints = 2 * abs(radius) * sin(angleBetweenWayPoints / 2);
            for (int i = 0; i < totalWayPoints; i++) {
                wayPoints.add(new TrackWayPoint(this, wayPointPosition, i, distanceFromStart + i));
                wayPointPosition = wayPointPosition.plus(new Polar(distanceBetweenWayPoints, headingToNextWayPoint));
                headingToNextWayPoint += signum(radius) * angleBetweenWayPoints;
            }
        }

        this.wayPoints = Collections.unmodifiableList(wayPoints);
    }

    public boolean isStraight() {
        return radius == 0;
    }

    public int indexOnTrack() {
        return indexOnTrack;
    }

    public Vector start() {
        return start;
    }

    public double length() {
        return length;
    }

    public double heading() {
        return heading;
    }

    public double radius() {
        return radius;
    }

    public double width() {
        return width;
    }

    public List<TrackWayPoint> wayPoints() {
        return wayPoints;
    }
}

File: src/main/kotlin/akostenko/aicars/track/TrackSection.kt
package akostenko.aicars.track

import akostenko.aicars.drawing.ArcLine
import akostenko.aicars.drawing.Line
import akostenko.aicars.drawing.StraightLinesBuilder
import akostenko.aicars.math.Polar
import akostenko.aicars.math.Vector
import java.lang.StrictMath.abs
import java.lang.StrictMath.signum
import java.lang.StrictMath.sin
import java.util.*

class TrackSection internal constructor(distanceFromStart: Int,
                                        val indexOnTrack: Int,
                                        val start: Vector,
                                        val length: Double,
                                        val radius: Double,
                                        val heading: Double // for turns it's a tangent line to the beginning point
                                        , val width: Double) {
    val wayPoints: List<TrackWayPoint>

    init {
        if (length <= 0) {
            throw IllegalArgumentException("Length of track section must be positive!")
        }

        val totalWayPoints = length.toInt() / wayPointStep + if (length % wayPointStep > 0) 1 else 0
        val wayPoints = ArrayList<TrackWayPoint>(totalWayPoints)
        var wayPointPosition = start
        var headingToNextWayPoint = heading

        if (radius == 0.0) {
            for (i in 0..totalWayPoints - 1) {
                wayPoints.add(TrackWayPoint(this, wayPointPosition, i, distanceFromStart + i))
                wayPointPosition = wayPointPosition.plus(Polar(wayPointStep.toDouble(), headingToNextWayPoint))
            }
        } else {
            val angleBetweenWayPoints = wayPointStep / abs(radius)
            val distanceBetweenWayPoints = 2.0 * abs(radius) * sin(angleBetweenWayPoints / 2)
            for (i in 0..totalWayPoints - 1) {
                wayPoints.add(TrackWayPoint(this, wayPointPosition, i, distanceFromStart + i))
                wayPointPosition = wayPointPosition.plus(Polar(distanceBetweenWayPoints, headingToNextWayPoint))
                headingToNextWayPoint += signum(radius) * angleBetweenWayPoints
            }
        }

        this.wayPoints = Collections.unmodifiableList(wayPoints)
    }

    val isStraight: Boolean = radius == 0.0

    val borders: Collection<Line> = if (isStraight) getStraightBorderLines() else getArcBorderLines()

    private fun getStraightBorderLines(): Collection<Line> {
        val sectionStart = start.toPolar()
        val sectionEnd = start.toPolar() + Polar(length, heading)

        val rightBorderOffset = Polar(width / 2, heading + Math.PI /2)
        val leftBorderOffset = Polar(width / 2, heading - Math.PI /2)
        return StraightLinesBuilder()
                .between(sectionStart + rightBorderOffset, sectionEnd + rightBorderOffset)
                .between(sectionStart +  leftBorderOffset, sectionEnd + leftBorderOffset)
                .build() +
                if (indexOnTrack == 0)
                    StraightLinesBuilder().between(sectionStart + leftBorderOffset, sectionStart + rightBorderOffset)
                            .build()
                    else emptyList()
    }

    private fun getArcBorderLines(): Collection<Line> {
        val center = start.toDecart() + Polar(radius, heading + Math.PI /2)
        var from = (start - center).toPolar().d - if (radius < 0) Math.PI else 0.0
        var to = from + length / radius
        if (radius < 0) {
            val t = to
            to = from
            from = t
        }
        return listOf(ArcLine(center, radius - width/2, from, to),
                ArcLine(center, radius + width/2, from, to)) +
                if (indexOnTrack == 0) {
                    val directionToStartFromCenter = Polar(1.0, (start - center).toPolar().d)
                    StraightLinesBuilder().between(center + directionToStartFromCenter*(radius-width/2),
                            center + directionToStartFromCenter*(radius+width/2))
                            .build()
                }
                else emptyList()
    }

    companion object {
        private val wayPointStep = 1 // m
    }
}

